File: typescript/systemdesign/circuitBreaker.ts
/*
Circuit breaker state machine (closed/open/half-open) wrapping a "remote
call" that can fail: trips open after a failure threshold, waits a cooldown,
then allows a limited trial in half-open before fully closing or re-opening.

status - completed
*/

import { logp } from "../utils/logger";

enum CircuitState {
  CLOSED = "CLOSED",
  OPEN = "OPEN",
  HALF_OPEN = "HALF_OPEN",
}

class CircuitOpenError extends Error {
  constructor() {
    super("circuit is open, call rejected");
  }
}

class CircuitBreaker {
  private state: CircuitState = CircuitState.CLOSED;
  private consecutiveFailures = 0;
  private openedAtMs = 0;

  constructor(
    private readonly failureThreshold: number,
    private readonly cooldownMs: number,
    private readonly nowFn: () => number = Date.now
  ) {}

  getState(): CircuitState {
    if (this.state === CircuitState.OPEN && this.nowFn() - this.openedAtMs >= this.cooldownMs) {
      this.state = CircuitState.HALF_OPEN;
    }
    return this.state;
  }

  call<T>(action: () => T): T {
    const currentState = this.getState();

    if (currentState === CircuitState.OPEN) {
      throw new CircuitOpenError();
    }

    try {
      const result = action();
      this.onSuccess();
      return result;
    } catch (err) {
      this.onFailure();
      throw err;
    }
  }

  private onSuccess(): void {
    this.consecutiveFailures = 0;
    this.state = CircuitState.CLOSED;
  }

  private onFailure(): void {
    this.consecutiveFailures++;
    if (this.state === CircuitState.HALF_OPEN || this.consecutiveFailures >= this.failureThreshold) {
      this.state = CircuitState.OPEN;
      this.openedAtMs = this.nowFn();
    }
  }
}

// usage scenarios: a flaky remote call plus a fake clock

let simulatedNow = 0;
let shouldFail = true;
const flakyRemoteCall = () => {
  if (shouldFail) {
    throw new Error("remote service unavailable");
  }
  return "ok";
};

const breaker = new CircuitBreaker(3, 1000, () => simulatedNow);

for (let i = 1; i <= 3; i++) {
  try {
    breaker.call(flakyRemoteCall);
  } catch (err) {
    logp(`call ${i} failed (${(err as Error).message}), state=${breaker.getState()}`);
  }
}

try {
  breaker.call(flakyRemoteCall);
} catch (err) {
  logp(`call 4 rejected immediately: ${(err as Error).message}, state=${breaker.getState()}`);
}

logp("advancing simulated clock past cooldown");
simulatedNow += 1000;
logp(`state after cooldown -> ${breaker.getState()} (expected HALF_OPEN)`);

shouldFail = false;
const result = breaker.call(flakyRemoteCall);
logp(`trial call in half-open succeeded -> "${result}", state=${breaker.getState()} (expected CLOSED)`);

File: python/systemdesign/circuit_breaker.py
"""
Circuit breaker state machine (closed / open / half-open) wrapping a
flaky downstream call. Trips to open after a run of consecutive
failures to stop hammering a failing dependency, waits out a cooldown,
then probes with a half-open trial before fully closing again.

status - completed
"""

import time

from base_logger.logging_event import create_logger


class CircuitOpenError(Exception):
    pass


class CircuitBreaker:

    log = create_logger(__name__)

    CLOSED = "CLOSED"
    OPEN = "OPEN"
    HALF_OPEN = "HALF_OPEN"

    def __init__(self, failure_threshold=3, cooldown_seconds=10, clock=time.monotonic):
        self.failure_threshold = failure_threshold
        self.cooldown_seconds = cooldown_seconds
        self.clock = clock
        self.state = self.CLOSED
        self.consecutive_failures = 0
        self.opened_at = None

    def _maybe_transition_to_half_open(self):
        if self.state == self.OPEN and self.clock() - self.opened_at >= self.cooldown_seconds:
            self.state = self.HALF_OPEN
            self.log.info("cooldown elapsed, transitioning OPEN -> HALF_OPEN")

    def call(self, func, *args, **kwargs):
        self._maybe_transition_to_half_open()

        if self.state == self.OPEN:
            raise CircuitOpenError("circuit is open, call rejected")

        try:
            result = func(*args, **kwargs)
        except Exception:
            self._on_failure()
            raise
        else:
            self._on_success()
            return result

    def _on_success(self):
        if self.state == self.HALF_OPEN:
            self.log.info("trial call succeeded, transitioning HALF_OPEN -> CLOSED")
        self.state = self.CLOSED
        self.consecutive_failures = 0
        self.opened_at = None

    def _on_failure(self):
        if self.state == self.HALF_OPEN:
            self.log.info("trial call failed, transitioning HALF_OPEN -> OPEN")
            self.state = self.OPEN
            self.opened_at = self.clock()
            return

        self.consecutive_failures += 1
        if self.consecutive_failures >= self.failure_threshold:
            self.log.info("failure threshold reached, transitioning CLOSED -> OPEN")
            self.state = self.OPEN
            self.opened_at = self.clock()


if __name__ == "__main__":
    log = create_logger("circuit_breaker_demo")

    fake_time = [0.0]

    def fake_clock():
        return fake_time[0]

    breaker = CircuitBreaker(failure_threshold=3, cooldown_seconds=5, clock=fake_clock)

    call_should_fail = [True]

    def flaky_dependency():
        if call_should_fail[0]:
            raise RuntimeError("downstream error")
        return "ok"

    for i in range(3):
        try:
            breaker.call(flaky_dependency)
        except RuntimeError:
            log.info("call %d failed, state=%s", i + 1, breaker.state)

    log.info("after 3 consecutive failures, state=%s (expected OPEN)", breaker.state)

    try:
        breaker.call(flaky_dependency)
    except CircuitOpenError:
        log.info("call rejected immediately while OPEN, dependency not invoked")

    fake_time[0] = 6.0
    call_should_fail[0] = False
    result = breaker.call(flaky_dependency)
    log.info("after cooldown, trial call result=%s, state=%s (expected CLOSED)",
             result, breaker.state)

    result = breaker.call(flaky_dependency)
    log.info("subsequent call result=%s, state=%s", result, breaker.state)
